### PR TITLE
[CARBONDATA-4274] Fix create partition table error with spark 3.1

### PR DESCRIPTION
**Pull request by ShreelekhyaG**
### Why is this PR needed?
 With spark 3.1, we can create a partition table by giving partition columns from schema. 
 Like below example:
`create table partitionTable(c1 int, c2 int, v1 string, v2 string) stored as carbondata partitioned by (v2,c2)`

When the table is created by SparkSession with CarbonExtension, catalog table is created with the specified partitions.
But in cluster/ with carbon session, when we create partition table with above syntax it is creating normal table with no partitions.
 
 ### What changes were proposed in this PR?
`partitionByStructFields` is empty when we directly give partition column names. So it was not creating a partition table. Made changes to identify the partition column names and get the struct field and datatype info from table columns.
    
 ### Does this PR introduce any user interface change?
 - No

 ### Is any new testcase added?
 - Yes, tested in cluster.

### CODE REVIEW

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/277/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5874/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/4132/

**Comment by MarvinLitt** [comment]
how does the CI pass under spark3.1?

**Comment by ShreelekhyaG** [comment]
> how does the CI pass under spark3.1?
> the partition table test case should work to take care of this issue, right?

Prior to this, there is no such test case in CI with syntax like below, where we can create a partition table by giving partition column names from schema.
`create table partitionTable(c1 int, c2 int, v1 string, v2 string) stored as carbondata partitioned by (v2,c2)`

With 2.3 and 2.4 versions, the above syntax will fail while parsing and only the syntax with partition column names and datatype is valid like below. 
`create table partitionTable(c1 int, v1 string) stored as carbondata partitioned by (v2 string,c2 int)`

spark 3.1 supports creation of partition table with both of the above syntax types.

File: integration/spark/src/main/spark3.1/org/apache/spark/sql/parser/CarbonSparkSqlParser.scala [hunk 1]
```diff
@@ -129,9 +129,32 @@ class CarbonHelperSqlAstBuilder(conf: SQLConf,
...
     // validate partition clause
-    val partitionByStructFields = Option(partitionColumns).toSeq
-        .flatMap(x => visitPartitionFieldList(x)._2)
+    var partitionTransformList: Seq[Transform] = Seq()
+    var partitionByStructFields: Seq[StructField] = Seq()
+    Option(partitionColumns).toSeq.foreach(x => {
+      val (t, s) = visitPartitionFieldList(x)
+      partitionTransformList = partitionTransformList ++ t
+      partitionByStructFields = partitionByStructFields ++ s
+    })
```
Comment (MarvinLitt) [inline integration/spark/src/main/spark3.1/org/apache/spark/sql/parser/CarbonSparkSqlParser.scala:141]: Option(partitionColumns).toSeq.foreach(x => {
      val (t, s) = visitPartitionFieldList(x)
      partitionTransformList = partitionTransformList ++ t
      partitionByStructFields = partitionByStructFields ++ s
    })  
does this no need to use foreach, the input of visitPartitionFieldList is PartitionFieldListContext type and after func visitPartitionFieldList the result is two list.

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/286/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5884/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/4141/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/288/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5886/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/4143/

**Comment by ShreelekhyaG** [comment]
retest this please

File: integration/spark/src/main/spark3.1/org/apache/spark/sql/parser/CarbonSparkSqlParser.scala (hunk 1 above)
Comment (ShreelekhyaG) [reply integration/spark/src/main/spark3.1/org/apache/spark/sql/parser/CarbonSparkSqlParser.scala]: Have made the change to directly call the `visitPartitionFieldList `function.

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/289/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5887/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/4144/

File: integration/spark/src/main/spark3.1/org/apache/spark/sql/parser/CarbonSparkSqlParser.scala [hunk 2]
```diff
@@ -130,8 +131,24 @@ class CarbonHelperSqlAstBuilder(conf: SQLConf,
...
-        .flatMap(x => visitPartitionFieldList(x)._2)
+    var (partitionTransformList,
+    partitionByStructFields) = visitPartitionFieldList(partitionColumns)
+    // If the struct fields is empty but partition columns are given,
+    // then search in table columns to extract the struct fields.
+    if (partitionByStructFields.isEmpty && !partitionTransformList.isEmpty) {
+      val partitionNames = partitionTransformList
+        .flatMap(_.references().flatMap(_.fieldNames()))
+      partitionNames.foreach(partName => {
+        val ans = cols.find(x => x.name.equals(partName))
```
Comment (kunal642) [inline integration/spark/src/main/spark3.1/org/apache/spark/sql/parser/CarbonSparkSqlParser.scala:142]: change var name to structField

File: integration/spark/src/main/spark3.1/org/apache/spark/sql/parser/CarbonSparkSqlParser.scala [hunk 3]
```diff
@@ -130,8 +131,24 @@ class CarbonHelperSqlAstBuilder(conf: SQLConf,
     val tableProperties = convertPropertiesToLowercase(properties)
 
     // validate partition clause
-    val partitionByStructFields = Option(partitionColumns).toSeq
-        .flatMap(x => visitPartitionFieldList(x)._2)
+    var (partitionTransformList,
+    partitionByStructFields) = visitPartitionFieldList(partitionColumns)
+    // If the struct fields is empty but partition columns are given,
+    // then search in table columns to extract the struct fields.
+    if (partitionByStructFields.isEmpty && !partitionTransformList.isEmpty) {
```
Comment (kunal642) [inline integration/spark/src/main/spark3.1/org/apache/spark/sql/parser/CarbonSparkSqlParser.scala:138]: change !Seq().empty to Seq().nonEmpty

File: integration/spark/src/main/spark3.1/org/apache/spark/sql/parser/CarbonSparkSqlParser.scala (hunk 2 above)
Comment (ShreelekhyaG) [reply integration/spark/src/main/spark3.1/org/apache/spark/sql/parser/CarbonSparkSqlParser.scala]: Done

File: integration/spark/src/main/spark3.1/org/apache/spark/sql/parser/CarbonSparkSqlParser.scala (hunk 3 above)
Comment (ShreelekhyaG) [reply integration/spark/src/main/spark3.1/org/apache/spark/sql/parser/CarbonSparkSqlParser.scala]: Done

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5893/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/4149/

**Comment by kunal642** [comment]
retest this please

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5903/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/303/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/4158/

**Pull request closed without merging**